### PR TITLE
fix custom op for backward compatibility

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description ##
The default implementation `infer_storage_type_backward` in custom op has some problem.

Before this PR, the LR in sparse example is broken due to the custom op, and the added test will throw err message,
```
Error in mult.infer_type: Traceback (most recent call last):
  File "/home/hanfeng/zyh/mxnet/python/mxnet/operator.py", line 737, in infer_storage_type_backward_entry
    "stypes, got %d."%(total_outputs, len(ostype))
AssertionError: InferStorageTypeBackward Error: expecting 2 entries in returned output stypes, got 1.

[10:27:12] /home/hanfeng/zyh/mxnet/dmlc-core/include/dmlc/./logging.h:308: [10:27:12] src/operator/custom/custom.cc:383: Check failed: reinterpret_cast<CustomOpBackwardInferStorageTypeFunc>( params.info->callbacks[kCustomOpPropBackwardInferStorageType])( stypes.size(), stypes.data(), params.info->contexts[kCustomOpPropBackwardInferStorageType])
```
cc @eric-haibin-lin @anirudh2290

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [ ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
@anirudh2290

**Comment by anirudh2290** [comment]
Thank you for fixing this and adding the tests! Yes, since the output of stypes which is the in_grads would depend on arguments length and not output length. Can you please let me know what LR example you are referring to?

File: tests/python/unittest/test_operator.py [hunk 1]
```diff
@@ -3652,6 +3652,42 @@ def create_operator(self, ctx, shapes, dtypes):
     assert (x.grad.stype == 'csr')
     assert (y.stype == 'csr')
     assert (aux.stype == 'csr')
+    
+    # test for backward compatibility, i.e. the correctness of default implementation of
```
Comment (anirudh2290) [inline tests/python/unittest/test_operator.py]: Please remove whitespaces here and above.

**Comment by ZiyueHuang** [comment]
@anirudh2290 Thanks for your comments. LR example is the `linear_classification.py` in example/sparse.

File: tests/python/unittest/test_operator.py [hunk 2]
```diff
@@ -3652,6 +3652,43 @@ def create_operator(self, ctx, shapes, dtypes):
...
+            self.assign(out_data[0], req[0], in_data[0]*in_data[1])
+
+        def backward(self, req, out_grad, in_data, out_data, in_grad, aux):
+            self.assign(in_grad[0], req[0], in_data[1]*out_grad[0])
+            self.assign(in_grad[1], req[1], in_data[0]*out_grad[0])
+
+    @mx.operator.register("mult")
+    class MultProp(mx.operator.CustomOpProp):
+        def __init__(self):
+            super(MultProp, self).__init__(need_top_grad=True)
```
Comment (anirudh2290) [inline tests/python/unittest/test_operator.py]: Found an issue with custom_op when used with need_top_grad=False. Fixing here: #8725

**Comment by piiswrong** [comment]
Closing this because we are reverting sparse support for custom op

**Pull request closed without merging**